File: examples/kotlin/features/effects-selection/kotlin/cott_impl/curriculum/effects_selection/copy_text.kt
package cott_impl.curriculum.effects_selection

internal fun copy_text(source: java.nio.file.Path, destination: java.nio.file.Path): cott_runtime.CottResult<curriculum.effects_selection.CopyReceipt, curriculum.effects_selection.EffectError> {
    val file = when (val result = curriculum.effects_selection.read_text(source)) {
        is cott_runtime.Ok -> result.value
        is cott_runtime.Err -> return cott_runtime.Err(result.error)
    }
    return try {
        val bytes = file.text.toByteArray(kotlin.text.Charsets.UTF_8)
        val target = destination.toAbsolutePath()
        val temporary = java.nio.file.Files.createTempFile(target.parent, ".cott-copy-", ".tmp")
        try {
            java.nio.file.Files.write(temporary, bytes)
            java.nio.file.Files.move(
                temporary,
                target,
                java.nio.file.StandardCopyOption.ATOMIC_MOVE,
                java.nio.file.StandardCopyOption.REPLACE_EXISTING
            )
        } catch (failure: Exception) {
            try {
                java.nio.file.Files.deleteIfExists(temporary)
            } catch (cleanupFailure: Exception) {
                failure.addSuppressed(cleanupFailure)
            }
            throw failure
        }
        cott_runtime.Ok(curriculum.effects_selection.CopyReceipt(destination, bytes.size.toULong()))
    } catch (failure: Exception) {
        cott_runtime.Err(
            curriculum.effects_selection.EffectError.OperationFailed(
                failure.message ?: "Failed to copy text"
            )
        )
    }
}

File: examples/features/effects-selection/python/_cott_impl/curriculum/effects_selection/copy_text.py
from pathlib import Path

from cott_runtime import CottContractViolation, Err, Ok, Result, _cott_fixture_replace
from curriculum.effects_selection import read_text
from curriculum.effects_selection_types import CopyReceipt, EffectError, EffectError_OperationFailed


def copy_text(source: Path, destination: Path) -> Result[CopyReceipt, EffectError]:
    match read_text(source):
        case Err(error=error):
            return Err(error=error)
        case Ok(value=file):
            data = file.text.encode("utf-8")
            try:
                _cott_fixture_replace(destination, data)
            except (OSError, CottContractViolation) as exc:
                return Err(error=EffectError_OperationFailed(message=str(exc)))
            return Ok(value=CopyReceipt(destination=destination, bytes_written=len(data)))
